ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -189,19 +188,19 @@ sealed abstract class Queries(tablePrefix: String, tpIdCacheMaxEntries: Long)(im
 
   private def surrogateTemplateIdFromDb(packageId: String, moduleName: String, entityName: String)(
       implicit log: LogHandler
-  ): ConnectionIO[SurrogateTpId] =
-    sql"""SELECT tpid FROM $templateIdTableName
-          WHERE (package_id = $packageId AND template_module_name = $moduleName
-                 AND template_entity_name = $entityName)"""
-      .query[SurrogateTpId]
-      .option flatMap {
-      _.cata(
-        _.pure[ConnectionIO],
-        sql"""INSERT INTO $templateIdTableName (package_id, template_module_name, template_entity_name)
-              VALUES ($packageId, $moduleName, $entityName)""".update
-          .withUniqueGeneratedKeys[SurrogateTpId]("tpid"),
-      )
-    }
+  ): ConnectionIO[SurrogateTpId] = for {
+    _ <- insertTemplateIdIfNotExists(packageId, moduleName, entityName).update.run
+    tpid <- sql"""SELECT tpid FROM $templateIdTableName
+                  WHERE (package_id = $packageId AND template_module_name = $moduleName
+                      AND template_entity_name = $entityName)""".query[SurrogateTpId].unique
```

Comment:
Is there still a race here - in that the read (SELECT) will be done in a subsequent transaction to the write (INSERT)? I am wondering if we have a fundamental auto-commit-based JDBC issue where we are assuming these things are executed transactionally, but they are not? The risk here is that the read does not return what was written.

Reply:
There can only ever be one row per with that `(packageId, moduleName, entityName)` triple, the table constraint ensures that, we just tell it to treat an attempt to INSERT a pre-existing value as a no-op. Then there can only be one value that the SELECT can return.

Reply:
Even if the SELECT is in subsequent transaction it doesn't matter because at that point you get the templateId no matter whether the templateId was inserted into DB by your thread or somebody else's thread.

Reply:
Yes, I am thinking about a concurrent operation - it is possible the value changes between the INSERT and the SELECT (?) Maybe that is not a potential scenario.

Reply:
No, it never gets changed or deleted.

Reply:
It's in the same transaction; `ConnectionIO` does not implicitly commit on flatmap. When you natural-transform `ConnectionIO` to another functor like `Future`, or explicitly perform the `fconn.commit` or certain other `ConnectionIO` transaction-actions, you change transaction.